packaging: add spec for Fedora and build-rpm script

cjp256 opened the pull request:
Add unofficial specs for Mariner and Fedora to build test RPMs from current git tree.

Tested on Mariner 2, Azure Linux 3.0, RHEL 9.3 and Fedora 40.

jeremycline commented on packaging/mariner/azure-nvme-utils.spec.tmpl:
```diff
@@ -1,11 +1,11 @@
-Version:        0.1.1
-Release:        1%{?dist}
+Version:        __VERSION__
+Release:        __RELEASE__%{?dist}
```
If you don't want to mess with the template you can achieve the same thing with something like:

```suggestion
Version:        %{__git_version}
Release:        %{__git_release}%{?dist}
```
And then alter your build call to
```
rpmbuild -ba --define "__git_version $version" --define "__git_release $release" azure-nvme-utils.spec
```

cjp256 replied:
works great, thanks! much simpler

jeremycline commented on packaging/mariner/azure-nvme-utils.spec.tmpl:
```diff
@@ -20,16 +20,15 @@ Utility and udev rules to help identify Azure NVMe devices.
 
 %files
-/usr/sbin/azure-nvme-id
 /lib/udev/rules.d/80-azure-nvme.rules
```
```suggestion
%{_libdir}/udev/rules.d/80-azure-nvme.rules
```

cjp256 replied:
libdir maps to /usr/lib64 on the Fedora image.... i don't see a macro for /usr/lib at https://docs.fedoraproject.org/en-US/packaging-guidelines/RPMMacros/

fixed other macros and updated this to %{_exec_prefix}/lib

jeremycline commented on packaging/mariner/azure-nvme-utils.spec.tmpl:
```diff
@@ -20,16 +20,15 @@ Utility and udev rules to help identify Azure NVMe devices.
-%changelog
-* Mon Feb 26 2024 Chris Patterson <cpatterson@microsoft.com> - 0.1.1-1
-- Initial package
+/usr/sbin/azure-nvme-id
```
```suggestion
%{_sbindir}/azure-nvme-id
```
FWIW https://fedoraproject.org/wiki/Changes/Unify_bin_and_sbin is coming so you could get ahead of the game and use `%{_bindir}` instead.

cjp256 replied:
I will do that in a separate PR to update it to bindir everywhere

jeremycline commented on scripts/build-rpm.sh:
```diff
@@ -0,0 +1,47 @@
+
+# Create spec file from template.
+cd "$project_dir/packaging/$distro"
+cat azure-nvme-utils.spec.tmpl | sed "s/__VERSION__/$version/g" | sed "s/__RELEASE__/$release/g" > azure-nvme-utils.spec
```
If you decide to not go the template route this doesn't matter, but ShellCheck notes this could be written
```suggestion
sed "s/__VERSION__/$version/g" azure-nvme-utils.spec.tmpl | sed "s/__RELEASE__/$release/g" > azure-nvme-utils.spec
```

cjp256 replied:
updated, thanks!

jeremycline commented on scripts/build-rpm.sh:
```diff
@@ -0,0 +1,47 @@
+mkdir -p "${build_dir}"/rpmbuild/{BUILD,RPMS,SOURCES,SPECS,SRPMS}
+
+# Azure Linux / Mariner requires rpmmacros.
+if [[ $ID == "mariner" || $ID == "azurelinux" ]]; then
```
Similar to the version and release macros I suggested above, you should be able to use `--define "_topdir ..."` when you build the RPM.

cjp256 replied:
great, thanks! done

jeremycline commented on packaging/fedora/azure-nvme-utils.spec.tmpl:
```diff
@@ -0,0 +1,34 @@
+
+License:        MIT
+URL:            https://github.com/Azure/%{name}
+Source0:        azure-nvme-utils_dev.tgz
```
I think this is fine, but if you wanted to get fancy you could use https://docs.fedoraproject.org/en-US/packaging-guidelines/SourceURL/#_using_forges_hosted_revision_control. Presumably you'll be tagging releases for the Fedora package so we'd need to adjust it regardless, so it's probably not worth it.

cjp256 replied:
yeah this will be done differently for proper Fedora packaging, we'll do whatever makes the most sense there when we get there.  I have a couple minor fixes I want to address here first then will start the upstream packaging process

jeremycline reviewed (commented):
Fedora does support [autorelease](https://docs.fedoraproject.org/en-US/packaging-guidelines/Versioning/#_release_tag) so I would recommend when packaging this for Fedora using that, and of course the Source needs be a URL (and you can use the forge macros for that if you want).

cjp256 committed:
packaging: add spec for Fedora/RHEL and build-rpm script

Add unofficial specs for Mariner and Fedora to build test RPMs from
current git tree.

Tested on Mariner 2, Fedora 40, RHEL 9.3 and Azure Linux 3.0.

Signed-off-by: Chris Patterson <cpatterson@microsoft.com>

jeremycline commented on packaging/mariner/azure-nvme-utils.spec:
```diff
@@ -20,16 +20,15 @@ Utility and udev rules to help identify Azure NVMe devices.
-- Initial package
+%files
+%{_libdir}/udev/rules.d/80-azure-nvme.rules
+%{_sbindir}/azure-nvme-id
```
Did you mean to include autochangelog after this bit?

cjp256 replied:
i set it for fedora/rhel but mariner doesnt support it

jeremycline replied:
Ah, fair enough. Did you mean to remove the non-autochangelog, or does that get added by some other mechanism in Mariner?

cjp256 replied:
yeah doesn't seem required and keeping a changelog pointless for this dev build

cjp256 replied:
fwiw here's the upstream packaging: https://github.com/microsoft/azurelinux/blob/3.0-dev/SPECS/azure-nvme-utils/azure-nvme-utils.spec#L33  a bit dated and will be updated to reflect the suggestions you made here after next tag

jeremycline replied:
Good point, I'd forgotten this was only for dev builds. I kinda expected RPM to require one but if it works this way, all the better.

jeremycline reviewed (commented):
Looks good to me, I think you might have missed including a changelog of some kind for the Mariner version?
